Expose altitudeAngle and azimuthAngle to PointerEvent

lilyspiniolas opened the pull request:
#### 1b6a1bb6dfcbbc09d6e224de75eb932a2e97c64d
<pre>
Expose altitudeAngle and azimuthAngle to PointerEvent
<a href="https://bugs.webkit.org/show_bug.cgi?id=276755">https://bugs.webkit.org/show_bug.cgi?id=276755</a>
<a href="https://rdar.apple.com/131974392">rdar://131974392</a>

Reviewed by Abrar Rahman Protyasha.

These changes make the altitudeAngle and azimuthAngle attributes accessible as part of the PointerEvent interface.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/PointerEvent.cpp:
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/dom/PointerEvent.h:
* Source/WebCore/dom/PointerEvent.idl:
* Source/WebCore/dom/ios/PointerEventIOS.cpp:
(WebCore::m_isPrimary):
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.mm:

Canonical link: <a href="https://commits.webkit.org/282017@main">https://commits.webkit.org/282017@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6c8f8d84386255549a6499ab73adb8cbe5c86480

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61635 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40988 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14226 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65609 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12180 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63754 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48673 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12451 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49727 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8453 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64704 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38083 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53388 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30560 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34742 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10618 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11111 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/54730 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56551 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10918 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67337 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60876 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5578 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10680 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57103 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5603 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53337 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57323 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13743 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4588 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/82637 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36789 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14446 "Found 1466 new JSC stress test failures: microbenchmarks/memcpy-wasm-large.js.no-cjit-validate-phases, microbenchmarks/memcpy-wasm-medium.js.no-cjit-validate-phases, microbenchmarks/memcpy-wasm-small.js.no-cjit-validate-phases, microbenchmarks/memcpy-wasm.js.no-cjit-validate-phases, microbenchmarks/wasm-cc-int-to-int.js.default-wasm, stress/proxy-get-and-set-recursion-stack-overflow.js.eager-jettison-no-cjit, stress/proxy-get-and-set-recursion-stack-overflow.js.no-cjit-collect-continuously, stress/proxy-get-and-set-recursion-stack-overflow.js.no-cjit-validate-phases, stress/proxy-set.js.bytecode-cache, stress/proxy-set.js.default ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37873 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38969 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37618 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/dom/PointerEvent.idl:
```diff
@@ -53,6 +55,8 @@ dictionary PointerEventInit : MouseEventInit {
     readonly attribute long tiltY;
     readonly attribute long twist;
+    readonly attribute double altitudeAngle;
+    readonly attribute double azimuthAngle;
```
Please put these behind a preference in `UnifiedWebPreferences.yaml`.

pxlcoder reviewed (commented):
Some comments on the commit message:

`Combined changes:` should be removed.

Bugzilla and radar should appear on their own lines immediately after the title (right now there is another line before them).

pxlcoder commented on Source/WebCore/dom/PointerEvent.h:
```diff
@@ -56,6 +56,8 @@ class PointerEvent final : public MouseEvent {
         long tiltX { 0 };
         long tiltY { 0 };
         long twist { 0 };
+        double altitudeAngle { piOverTwoDouble / 2 };
```
`piOverFourDouble`?

aprotyas replied:
We just want `piOverTwoDouble` here, actually.

pxlcoder commented on Source/WebCore/dom/PointerEvent.h:
```diff
@@ -158,6 +162,8 @@ class PointerEvent final : public MouseEvent {
     long m_tiltX { 0 };
     long m_tiltY { 0 };
     long m_twist { 0 };
+    double m_altitudeAngle { piOverTwoDouble };
```
Is it expected that this differs from the value above?

pxlcoder commented on Source/WebCore/dom/ios/PointerEventIOS.cpp:
```diff
@@ -72,13 +72,10 @@ PointerEvent::PointerEvent(const AtomString& type, const PlatformTouchEvent& eve
-    m_tiltX = round(cos(azimuthAngle) * cos(altitudeAngle) * 90);
-    m_tiltY = round(sin(azimuthAngle) * cos(altitudeAngle) * 90);
+    m_azimuthAngle = event.azimuthAngleAtIndex(index);
+    m_altitudeAngle = m_pointerType == penPointerEventType() ? event.altitudeAngleAtIndex(index) : piOverTwoDouble;
```
The introduction of a `m_pointerType != penPointerEventType()` case is a behavior change for `m_tiltX` and `m_tiltY`. Is that expected?

cc @aprotyas

lilyspiniolas replied:
I believe so, at the moment tiltY returns 90 by default when it should return 0. This changes makes it so that both tiltX and tiltY are 0 by default

aprotyas replied:
Hmm, I'm not following along, why do we specialize based on the pointer type here?

pxlcoder reviewed (commented):
There will probably be some WPTs to rebaseline?

aprotyas commented on Source/WebCore/dom/PointerEvent.idl:
```diff
@@ -32,6 +32,8 @@ dictionary PointerEventInit : MouseEventInit {
     long tiltY = 0;
     long twist = 0;
+    double altitudeAngle;
+    double azimuthAngle;
```
You also want to gate these behind `[EnabledBySetting=...]`

lilyspiniolas replied:
If the pointer is a penPointerEventType() then a device that supports tilt (like the Apple Pencil) is being used so it can report a calculated value, if it is a different pointer then tilt isn't supported in which case it returns the default value.

aprotyas replied:
You probably meant to put this reply in the other thread!

aprotyas replied:
You ought to still gate those dict values behind `[EnabledBySetting=...]`?

aprotyas reviewed (commented):
It looks like [imported/w3c/web-platform-tests/pointerevents/pointerevent_tiltX_tiltY_to_azimuth_altitude.html](https://ews-build.s3-us-west-2.amazonaws.com/macOS-Sonoma-Debug-WK2-Tests-EWS/53fd282d-9548/imported/w3c/web-platform-tests/pointerevents/pointerevent_tiltX_tiltY_to_azimuth_altitude-pretty-diff.html) is a valid failure.

pxlcoder commented on Source/WebCore/dom/PointerEvent.cpp:
```diff
@@ -86,6 +96,54 @@ PointerEvent::PointerEvent()
+
+    else {
+        azimuthAngle = atan2(tan(tiltYRadians), tan(tiltXRadians));
+        azimuthAngle = azimuthAngle < 0 ? azimuthAngle + 2 * piDouble : azimuthAngle;
```
```suggestion
        azimuthAngle = azimuthAngle < 0 ? azimuthAngle + radiansPerTurnDouble : azimuthAngle;
```

pxlcoder commented on Source/WebCore/dom/PointerEvent.cpp:
```diff
@@ -86,6 +96,54 @@ PointerEvent::PointerEvent()
+// This algorithm is equivalent to the one provided in https://w3c.github.io/pointerevents/#converting-between-tiltx-tilty-and-altitudeangle-azimuthangle.
+static PointerEventTilt tiltFromAngle(double altitudeAngle, double azimuthAngle)
+{
+    const double radiansToDegrees = 180 / piDouble;
```
We have `degreesPerRadianDouble` for this.

hortont424 commented on Source/WebCore/dom/PointerEvent.cpp:
```diff
@@ -86,6 +96,54 @@ PointerEvent::PointerEvent()
+
+    else if (!tiltX)
+        altitudeAngle = piOverTwoDouble - abs(tiltYRadians);
+
```
these newlines between else cases are unusual

pxlcoder commented on Source/WebCore/dom/PointerEvent.cpp:
```diff
@@ -86,6 +96,54 @@ PointerEvent::PointerEvent()
+static PointerEventAngle angleFromTilt(long tiltX, long tiltY)
+{
+    const double tiltXRadians = tiltX * piDouble / 180;
+    const double tiltYRadians = tiltY * piDouble / 180;
```
```suggestion
    const double tiltXRadians = tiltX * radiansPerDegreeDouble;
    const double tiltYRadians = tiltY * radiansPerDegreeDouble;
```

hortont424 commented on Source/WebCore/dom/PointerEvent.cpp:
```diff
@@ -86,6 +96,54 @@ PointerEvent::PointerEvent()
+static PointerEventTilt tiltFromAngle(double altitudeAngle, double azimuthAngle)
+{
+    const double radiansToDegrees = 180 / piDouble;
+    double tiltXRadians = !altitudeAngle ? cos(azimuthAngle) * piOverTwoDouble : atan(cos(azimuthAngle) / tan(altitudeAngle));
```
should this ! be an areEssentiallyEqual(altitudeAngle, 0)? Floating point equality is fun.

lilyspiniolas replied:
Using areEssentiallyEqual seems to make the relevant pointer event test fail.

aprotyas commented on Source/WebCore/dom/PointerEvent.cpp:
```diff
@@ -36,6 +36,16 @@ namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(PointerEvent);
 
+typedef struct PointerEventTilt {
+    long tiltX;
+    long tiltY;
+} PointerEventTilt;
+
+typedef struct PointerEventAngle {
+    double altitudeAngle;
+    double azimuthAngle;
+} PointerEventAngle;
```
Why the `typedef struct` construct? This would be equivalent with less typing!

```suggestion
struct PointerEventTilt {
    long tiltX;
    long tiltY;
};

struct PointerEventAngle {
    double altitudeAngle;
    double azimuthAngle;
};
```

aprotyas commented on Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.mm:
```diff
@@ -339,7 +339,7 @@ - (void)_recordTouches:(NSSet<UITouch *> *)touches type:(WebKit::WKTouchEventTyp
         } else {
             touchPoint.touchType = WebKit::WKTouchPointType::Direct;
-            touchPoint.altitudeAngle = 0;
+            touchPoint.altitudeAngle = piOverTwoDouble;
```
This default value is non-obvious. Maybe make a remark in a comment that this is prescribed by the Pointer Events spec? Or even link to the spec text if you feel that's appropriate https://w3c.github.io/pointerevents/#dom-pointerevent-altitudeangle:~:text=angle%2C%20the%20value-,MUST%20be%20%CF%80/2,-.

aprotyas reviewed (approved):
:shipit: :shipit:

lilyspiniolas committed:
Expose altitudeAngle and azimuthAngle to PointerEvent
https://bugs.webkit.org/show_bug.cgi?id=276755
rdar://131974392

Reviewed by Abrar Rahman Protyasha.

These changes make the altitudeAngle and azimuthAngle attributes accessible as part of the PointerEvent interface.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/PointerEvent.cpp:
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/dom/PointerEvent.h:
* Source/WebCore/dom/PointerEvent.idl:
* Source/WebCore/dom/ios/PointerEventIOS.cpp:
(WebCore::m_isPrimary):
* Source/WebKit/UIProcess/ios/WKTouchEventsGestureRecognizer.mm:

Canonical link: https://commits.webkit.org/282017@main
